File: SnakeClient/GameClass/Snake.cs
using Microsoft.JSInterop;
using System.Text.Json;

namespace SnakeClient.GameClass
{
    public class Snake : AcGameObject
    {
        public int Id { get; set; }
        public string Color { get; set; }
        public GameMap GameMap { get; set; }
        public List<Cell> Cells { get; set; }
        public double Speed { get; set; } = 0.1;
        public int Direction { get; set; } = -1; // 0上 1右 2下 3左
        public string Status { get; set; } = "idle"; // idle, move, die
        public Cell? NextCell { get; set; }
        public int Step { get; set; } = 0;
        public int EyeDirection { get; set; }

        private readonly int[] dr = { -1, 0, 1, 0 };
        private readonly int[] dc = { 0, 1, 0, -1 };
        private readonly double eps = 1e-2;

        // 眼睛位置偏移
        private readonly double[,] eyeDx = new double[,] { { -1, 1 }, { 1, 1 }, { 1, -1 }, { -1, -1 } };
        private readonly double[,] eyeDy = new double[,] { { -1, -1 }, { -1, 1 }, { 1, 1 }, { 1, -1 } };

        public Snake(int id, string color, int r, int c, GameMap gameMap, GameLoop gameLoop) : base(gameLoop)
        {
            Id = id;
            Color = color;
            GameMap = gameMap;
            Cells = new List<Cell> { new Cell(r, c) };
            EyeDirection = id == 1 ? 2 : 0;
        }

        public bool CheckTailIncreasing()
        {
            return Step % 3 == 1;
        }

        public void SetDirection(int d)
        {
            Direction = d;
        }

        public void NextStep()
        {
            var d = Direction;
            NextCell = new Cell(Cells[0].R + dr[d], Cells[0].C + dc[d]);
            EyeDirection = d;
            Direction = -1;
            Status = "move";
            Step++;

            // 完全模拟JavaScript的逻辑
            var len = Cells.Count;
            Cells.Add(default!); // 添加占位元素，扩展 List 长度

            for (int i = len; i > 0; i--)
            {
                Cells[i] = DeepClone(Cells[i - 1]);
            }
        }

        public static T DeepClone<T>(T obj)
        {
            var json = System.Text.Json.JsonSerializer.Serialize(obj);
            return System.Text.Json.JsonSerializer.Deserialize<T>(json)!;
        }

        public override async Task UpdateAsync()
        {
            if (Status == "move")
            {
                UpdateMove();
            }
            await RenderAsync();
            Console.WriteLine(Cells.Count.ToString());
        }

        private void UpdateMove()
        {
            if (NextCell == null) return;

            var dx = NextCell.X - Cells[0].X;
            var dy = NextCell.Y - Cells[0].Y;
            var distance = Math.Sqrt(dx * dx + dy * dy);

            if (distance < eps)
            {
                //Console.WriteLine("++++++++++++++移动结束+++++++++++++++++");
                Cells[0] = NextCell;
                NextCell = null;
                Status = "idle";

                if (!CheckTailIncreasing())
                {
                    Cells.RemoveAt(Cells.Count - 1);
                }
            }
            else
            {
                //Console.WriteLine("===============正在移动===================");
                var moveDistance = Speed * timeDelta / 1000;
                //Console.WriteLine($"==============={moveDistance}===================");
                //Console.WriteLine($"==============={distance}===================");
                Cells[0].X += moveDistance * dx / distance;
                Cells[0].Y += moveDistance * dy / distance;

                if (!CheckTailIncreasing())
                {
                    var len = Cells.Count;
                    var tail = Cells[len - 1];
                    var tailTarget = Cells[len - 2];
                    var tailDx = tailTarget.X - tail.X;
                    var tailDy = tailTarget.Y - tail.Y;
                    var tailDistance = Math.Sqrt(tailDx * tailDx + tailDy * tailDy);
                    if (tailDistance > eps)
                    {
                        tail.X += moveDistance * tailDx / tailDistance;
                        tail.Y += moveDistance * tailDy / tailDistance;
                    }
                }
            }
        }

        private async Task RenderAsync()
        {
            var L = GameMap.L;
            var ctx = GameMap.Context;

            if (ctx == null) return;

            // 设置蛇身颜色
            await ctx.SetFillStyleAsync(Status == "die" ? "white" : Color);

            // 绘制蛇身圆圈
            foreach (var cell in Cells)
            {
                await ctx.BeginPathAsync();
                await ctx.ArcAsync(cell.X * L, cell.Y * L, L / 2 * 0.8, 0, Math.PI * 2);
                await ctx.FillAsync();
            }

            // 绘制蛇身连接部分
            for (int i = 1; i < Cells.Count; i++)
            {
                var a = Cells[i - 1];
                var b = Cells[i];

                if (Math.Abs(a.X - b.X) < eps && Math.Abs(a.Y - b.Y) < eps)
                    continue;
                else if (Math.Abs(a.X - b.X) < eps)
                    await ctx.FillRectAsync((a.X - 0.4) * L, Math.Min(a.Y, b.Y) * L, L * 0.8, Math.Abs(a.Y - b.Y) * L);
                else
                    await ctx.FillRectAsync(Math.Min(a.X, b.X) * L, (a.Y - 0.4) * L, Math.Abs(a.X - b.X) * L, L * 0.8);
            }

            // 绘制眼睛
            await ctx.SetFillStyleAsync("black");
            for (int i = 0; i < 2; i++)
            {
                var eyeX = (Cells[0].X + eyeDx[EyeDirection, i] * 0.15) * L;
                var eyeY = (Cells[0].Y + eyeDy[EyeDirection, i] * 0.15) * L;
                await ctx.BeginPathAsync();
                await ctx.ArcAsync(eyeX, eyeY, L * 0.065, 0, Math.PI * 2);
                await ctx.FillAsync();
            }
        }
    }
}

File: SnakeClient/wwwroot/Scripts/Snake.js
﻿import { AcGameObject } from "./AcGameObject";
import { Cell } from "./Cell";

export class Snake extends AcGameObject {
    constructor(info, gamemap) {
        super()

        this.id = info.id
        this.color = info.color
        this.gamemap = gamemap

        this.cells = [new Cell(info.r, info.c)] // 蛇的身体，由一个个Cell组成
        this.speed = 5
        this.direction = -1  // 0 1 2 3上右下左
        this.status = "idle"  // idle, move, die

        this.next_cell = null
        this.dr = [-1, 0, 1, 0]
        this.dc = [0, 1, 0, -1]
        this.step = 0   // 回合数

        this.eps = 1e-2

        this.eye_direction = 0
        if (this.id == 1) this.eye_direction = 2
        this.eye_dx = [
            [-1, 1],
            [1, 1],
            [1, -1],
            [-1, -1]
        ]
        this.eye_dy = [
            [-1, -1],
            [-1, 1],
            [1, 1],
            [1, -1]
        ]
    }

    check_tail_increasing() {  // 检测蛇长度增加
        //if(this.step <= 10)return true
        if (this.step % 3 === 1) return true
        return false
    }

    set_direction(d) {
        this.direction = d
    }

    start() {

    }
    next_step() {   //更新蛇的状态到下一步
        const d = this.direction

        this.next_cell = new Cell(this.cells[0].r + this.dr[d], this.cells[0].c + this.dc[d])
        this.eye_direction = d
        this.direction = -1
        this.status = "move"
        this.step++

        const len = this.cells.length
        for (let i = len; i > 0; i--) {
            this.cells[i] = JSON.parse(JSON.stringify(this.cells[i - 1]))
        }

        // if(!this.gamemap.check_valid(this.next_cell)){   // 目标格子会发生碰撞
        //     this.status = "die"
        //     return
        // }
        //
    }
    update_move() {
        const dx = this.next_cell.x - this.cells[0].x, dy = this.next_cell.y - this.cells[0].y
        const distance = Math.sqrt(dx * dx + dy * dy)
        if (distance < this.eps) {   // 到达目的地
            this.cells[0] = this.next_cell
            this.next_cell = null
            this.status = "idle"

            if (!this.check_tail_increasing()) {
                this.cells.pop()   //  蛇长不变，要去掉蛇尾
            }
        } else {
            const move_distance = this.speed * this.timedelta / 1000 // 每帧距离
            this.cells[0].x += move_distance * dx / distance
            this.cells[0].y += move_distance * dy / distance
            if (!this.check_tail_increasing()) {
                const len = this.cells.length
                const tail = this.cells[len - 1], tail_target = this.cells[len - 2]
                const tail_dx = tail_target.x - tail.x, tail_dy = tail_target.y - tail.y
                tail.x += move_distance * tail_dx / distance
                tail.y += move_distance * tail_dy / distance
            }
        }
    }
    update() {
        if (this.status === 'move') {
            this.update_move()
        }
        this.render()
    }
    render() {
        const L = this.gamemap.L
        const ctx = this.gamemap.ctx

        ctx.fillStyle = this.color
        if (this.status == "die") {
            ctx.fillStyle = "white"
        }

        for (const cell of this.cells) {
            ctx.beginPath()
            ctx.arc(cell.x * L, cell.y * L, L / 2 * 0.8, 0, Math.PI * 2)
            ctx.fill()
        }
        for (let i = 1; i < this.cells.length; i++) {
            const a = this.cells[i - 1], b = this.cells[i]
            if (Math.abs(a.x - b.x) < this.eps && Math.abs(a.y - b.y) < this.eps) {
                continue
            } else if (Math.abs(a.x - b.x) < this.eps) {
                ctx.fillRect((a.x - 0.4) * L, Math.min(a.y, b.y) * L, L * 0.8, Math.abs(a.y - b.y) * L)
            } else {
                ctx.fillRect(Math.min(a.x, b.x) * L, (a.y - 0.4) * L, Math.abs(a.x - b.x) * L, L * 0.8)
            }
        }
        ctx.fillStyle = 'black'
        for (let i = 0; i < 2; i++) {
            const eye_x = (this.cells[0].x + this.eye_dx[this.eye_direction][i] * 0.15) * L
            const eye_y = (this.cells[0].y + this.eye_dy[this.eye_direction][i] * 0.15) * L
            ctx.beginPath()
            ctx.arc(eye_x, eye_y, L * 0.065, 0, Math.PI * 2)
            ctx.fill()
        }
    }
}